proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,160 @@
+like `StructuredBuffer` may have RETs that exceed 128 bits, for example, if the target 
+IR is SPIR-V. The TypedBuffer rules above are only enforced when the IR target is DXIL.
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not an int, uint, nor a floating-point type, or vectors 
+or matrices containing the aforementioned types, will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for TypedBuffers.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom builtins to use as type_traits, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every TypedBuffer resource type that is defined, so that all
+TypedBuffer HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are RawBuffer variants, such as `StructuredBuffer`. These
+resource types will have a different set of type-traits applied, which will
+loosen constraints on viable RETs.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. The 
+preprocessor will take responsibility for selecting the right set of type_traits
+to check, depending on the target IR. If DXIL is given, then all of the TypedBuffer
+`type_traits` will be applied on each TypedBuffer HLSL resource type. Otherwise, the
+RawBuffer type_traits will be applied to each resource type. For every `type_trait`
+that is not true for the given RET, an associated error message will be emitted.
+
+The list of type_traits that will be available for use are described below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_resource_element_type` | An RET should be an arithmetic type, or a bool, or a vector or matrix or UDT containing such types. |

```

Comment:
What is different between this and `!__builtin_hlsl_is_intangible`?

Reply:
For the purposes of the single static assert. I don't think that just because an RET is not intangible, it is always allowed as an RET. Testing the affirmative seems to be better for long term. For example, what if HLSL supports strings in the future? We'd want to disallow strings as RETs.

Reply:
I'm not sure I agree about strings, but I'm also not sure how HLSL strings will be implemented, but my guess is that they'll be constant only so an offset and a length is the most I expect to represent the "value" of a string.

Setting aside strings since they aren't designed. As described, `__is_resource_element_type` is equivalent to `!__builtin_hlsl_is_intangible`, so I think we should have one trait not two because it simplifies the language rules.

Reply:
Talked with Chris offline, we both agree that `an arithmetic type, or a bool, or a vector or matrix or UDT containing such types` is exactly equivalent to `!__builtin_hlsl_is_intangible`, so I'll use that existing builtin.